File: packages/wasmvm/wasmlib/ts/wasmlib/dict.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {log} from './sandbox';
import {
    IKvStore,
    Proxy,
    stringFromBytes,
    WasmDecoder,
    WasmEncoder
} from './wasmtypes';

// returns a hex string representing the byte buffer
function hex(buf: Uint8Array): string {
    const hexa = '0123456789abcdef';
    let res = '';
    for (let i = 0; i < buf.length; i++) {
        const b = buf[i];
        res += hexa.charAt(b >> 4) + hexa.charAt(b & 0x0f);
    }
    return res;
}

function keya(key: Uint8Array): string {
    for (let i = 0; i < key.length; i++) {
        if (key[i] == 0x23) {
            return stringFromBytes(key.slice(0, i + 1)) + hex(key.slice(i + 1));
        }
        if (key[i] < 0x20 || key[i] > 0x7e) {
            return hex(key);
        }
    }
    return stringFromBytes(key);
}

function vala(val: Uint8Array): string {
    return hex(val);
}

export class ScDict implements IKvStore {
    dict: Map<string, Uint8Array> = new Map();

    public constructor(buf: Uint8Array | null) {
        if (buf !== null && buf.length != 0) {
            const dec = new WasmDecoder(buf);
           let size = dec.vluDecode(32) as u32;
            for (; size > 0; size--) {
                const keyLen = dec.vluDecode(32);
                const key = dec.fixedBytes(keyLen);
                const valLen = dec.vluDecode(32);
                const val = dec.fixedBytes(valLen);
                this.set(key, val);
            }
        }
    }

    static toKey(buf: Uint8Array): string {
        let key = '';
        for (let i = 0; i < buf.length; i++) {
            key += String.fromCharCode((buf[i] >> 4) + 0x40, (buf[i] & 0x0f) + 0x40);
        }
        return key;
    }

    static fromKey(key: string): Uint8Array {
        const buf = new Uint8Array(key.length / 2);
        for (let i = 0; i < key.length; i += 2) {
            const b1 = key.charCodeAt(i) as u8;
            const b2 = key.charCodeAt(i + 1) as u8;
            buf[i / 2] = (((b1 - 0x40) << 4) | (b2 - 0x40));
        }
        return buf;
    }

    public asProxy(): Proxy {
        return new Proxy(this);
    }

    delete(key: Uint8Array): void {
        // this.dump('delete');
        // log('dict.delete(' + keya(key) + ')');
        this.dict.delete(ScDict.toKey(key));
        // this.dump('Delete')
    }

    exists(key: Uint8Array): bool {
        const mapKey = ScDict.toKey(key);
        const ret = this.dict.has(mapKey);
        // this.dump('exists');
        // log('dict.exists(' + keya(key) + ') = ' + ret.toString());
        return ret;
    }

    get(key: Uint8Array): Uint8Array {
        // this.dump('get')
        const mapKey = ScDict.toKey(key);
        if (!this.dict.has(mapKey)) {
            // log('dict.get(' + keya(key) + ') = null');
            return new Uint8Array(0);
        }
        const value = this.dict.get(mapKey)!;
        // log('dict.get(' + keya(key) + ') = ' + vala(value));
        return value;
    }

    public immutable(): ScImmutableDict {
        return new ScImmutableDict(this);
    }

    set(key: Uint8Array, value: Uint8Array): void {
        // this.dump('set')
        // log('dict.set(' + keya(key) + ', ' + vala(value) + ')');
        this.dict.set(ScDict.toKey(key), value);
        // this.dump('Set')
    }

    public toBytes(): Uint8Array {
        if (this.dict.size == 0) {
            return new Uint8Array(1);
        }
        const keys = [...this.dict.keys()].sort();
        const enc = new WasmEncoder();
        enc.vluEncode(keys.length as u32);
        for (let i = 0; i < keys.length; i++) {
            const k = keys[i];
            const key = ScDict.fromKey(k);
            const val = this.dict.get(k)!;
            enc.vluEncode(key.length as u32);
            enc.fixedBytes(key, key.length as u32);
            enc.vluEncode(val.length as u32);
            enc.fixedBytes(val, val.length as u32);
        }
        return enc.buf();
    }

    protected dump(which: string): void {
        const keys = [...this.dict.keys()];
        for (let i = 0; i < keys.length; i++) {
            log('dict.' + which + '.' + i.toString() + '.' + keya(ScDict.fromKey(keys[i])) + ' = ' + vala(this.dict.get(keys[i])!));
        }
    }
}

export class ScImmutableDict {
    dict: Map<string, Uint8Array>;

    public constructor(dict: ScDict) {
        this.dict = dict.dict;
    }

    exists(key: Uint8Array): bool {
        return this.dict.has(ScDict.toKey(key));
    }

    get(key: Uint8Array): Uint8Array {
        const mapKey = ScDict.toKey(key);
        if (!this.dict.has(mapKey)) {
            return new Uint8Array(0);
        }
        return this.dict.get(mapKey)!;
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/dict.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"sort"

	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type ScImmutableDict struct {
	dict map[string][]byte
}

func (d *ScImmutableDict) Exists(key []byte) bool {
	return d.dict[string(key)] != nil
}

func (d *ScImmutableDict) Get(key []byte) []byte {
	return d.dict[string(key)]
}

type ScDict struct {
	ScImmutableDict
}

var _ wasmtypes.IKvStore = new(ScDict)

func NewScDict() *ScDict {
	return &ScDict{ScImmutableDict{dict: make(map[string][]byte)}}
}

func ScDictFromBytes(buf []byte) *ScDict {
	if len(buf) == 0 {
		return NewScDict()
	}
	dec := wasmtypes.NewWasmDecoder(buf)
	size := uint32(dec.VluDecode(32))
	dict := NewScDict()
	for i := uint32(0); i < size; i++ {
		keyLen := dec.VluDecode(32)
		key := dec.FixedBytes(uint32(keyLen))
		valLen := dec.VluDecode(32)
		val := dec.FixedBytes(uint32(valLen))
		dict.Set(key, val)
	}
	return dict
}

func (d *ScDict) AsProxy() wasmtypes.Proxy {
	return wasmtypes.NewProxy(d)
}

func (d *ScDict) Bytes() []byte {
	if d == nil {
		return []byte{0}
	}
	keys := make([]string, 0, len(d.dict))
	for key := range d.dict {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool {
		return keys[i] < keys[j]
	})
	enc := wasmtypes.NewWasmEncoder()
	enc.VluEncode(uint64(len(keys)))
	for _, k := range keys {
		key := []byte(k)
		val := d.dict[k]
		enc.VluEncode(uint64(len(key)))
		enc.FixedBytes(key, uint32(len(key)))
		enc.VluEncode(uint64(len(val)))
		enc.FixedBytes(val, uint32(len(val)))
	}
	return enc.Buf()
}

func (d *ScDict) Delete(key []byte) {
	delete(d.dict, string(key))
}

func (d *ScDict) Immutable() *ScImmutableDict {
	return &d.ScImmutableDict
}

func (d *ScDict) Set(key, value []byte) {
	d.dict[string(key)] = value
}
